language-plutus-core/src/Language/PlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -332,75 +363,87 @@ and transition it into
+   first evaluate each M_i to a value V_i and then call B on V_1, ...,
+   V_n.  We do this using FrameApplyBuiltin frames.  These contain a
+   list of the values V_1, ..., V_k which have alrady been evaluated
+   and a list of yet-to-be-evaluated term arguments.  When we finally run
+   out of term arguments, the function is called with the evaluated
+   arguments.
+
+   As pointed out in Note [Saved mapping example], we can't pass
+   arbitrary values to built-ins.  To deal with this, the
+   FrameApplyBuiltin frame also contains two environments: initialEnv,
+   which is the environment in effect when evaluation of B M_1 ... M_n
+   begin and finalEnv, which is to be used to carry on with the
+   computation when the builtin has been evaluated.  As we process the
+   arguments, we evaluate (in initialEnv) each one to a value V and
+   check whether that is a constant or not.  If V is a constant, we
+   save it directly in the list of arguments which will eventually be
+   fed to the builtin.  If V isn't a constant, we generate a fresh
+   variable v and extend finalEnv with a binding of v to a closure (V,
+   env) (where env is the current environment) and save v in the list of
+   arguments.
+
+   Thus when we finally come to evaluate B, each argument will either be
+   a constant or a variable, and finalEnv will be initialEnv extended
+   with bindings for the variables.  We apply B to these arguments to
+   obtain a result (possibly containing some of the input variables),
+   and continue execution in finalEnv, which contains the values
+   associated with all of the new variables.
+
+   Currently each argument is evaluated in initialEnv.  However, we might
+   be able to get away with using finalEnv instead, since every variable
+   in finalEnv\initialEnv is fresh and thus should not appear in any of the
+   input terms.
+
+   Note also that because builtin applications have to be saturated
+   with respect to both terms and types, a builtin cannot be
+   instantiated (partially or fully) separately from application: all
+   type parameters are supplied precisely at the point where the
+   builtin is invoked.  Thus (in contrast to the situation with
+   abstractions) instantiation of builtins cannot have any
+   computational effect.  FrameApplyBuiltin frames contain the type
+   parameters, but these are only used when generating a term for
+   error messages.  We could possibly dispense with them altogether.
+
+-}
+
+applyBuiltin :: (GShow uni, GEq uni, DefaultUni <: uni, Closed uni, uni `Everywhere` ExMemoryUsage)
+                 => VarEnv uni
+                 -> Context uni
+                 -> BuiltinName
+                 -> [Type TyName uni ExMemory]
+                 -> [Value TyName Name uni ExMemory]
+                 -> CekM uni (Term TyName Name uni ())
+applyBuiltin finalEnv ctx bn tys args =
+    -- If there are too many/few types this should be caught by the typechecker.
```

Comment:
Rather, by the builtin application machinery (from `Language.PlutusCore.Constant.Apply`), because we won't run the type checker on the chain.